---
title: 'Create Dataset'
output:
  html_document:
    code_folding: 'hide'
---

### Changes: Performed clustering using a signed hybrid network instead of unsigned because of [this post](https://peterlangfelder.com/2018/11/25/signed-or-unsigned-which-network-type-is-preferable/)


```{r load_packages, warning=FALSE, message=FALSE}
#setwd('/afs/inf.ed.ac.uk/user/s17/s1725186/Documents/PhD-Models/FirstPUModel/RMarkdowns')

library(tidyverse) ; library(reshape2) ; library(glue) ; library(plotly) ; library(dendextend)
library(RColorBrewer) ; library(viridis) ; require(gridExtra) ; library(GGally)
suppressWarnings(suppressMessages(library(WGCNA)))
library(expss)
library(polycor)
library(foreach) ; library(doParallel)
```

Load preprocessed dataset (preprocessing code in 19_10_14_data_preprocessing.Rmd) and clustering (pipeline in 19_10_21_WGCNA.Rmd)
```{r load_dataset, message=FALSE, warning=FALSE}

# Gandal dataset
load('./../Data/Gandal/preprocessed_data.RData')
datExpr = datExpr %>% data.frame
DE_info = DE_info %>% data.frame


# GO Neuronal annotations
GO_annotations = read.csv('./../../../InitialExperiments/FirstYearReview/Data/GO_annotations/genes_GO_annotations.csv')
GO_neuronal = GO_annotations %>% filter(grepl('neuron', go_term)) %>% 
              mutate('ID'=as.character(ensembl_gene_id)) %>% 
              dplyr::select(-ensembl_gene_id) %>% distinct(ID) %>%
              mutate('Neuronal'=1)


# SFARI Genes
SFARI_genes = read_csv('./../Data/SFARI/SFARI_genes_08-29-2019_with_ensembl_IDs.csv')
SFARI_genes = SFARI_genes[!duplicated(SFARI_genes$ID) & !is.na(SFARI_genes$ID),]


# Clusterings
clusterings = read_csv('./../Data/Gandal/clusters.csv')


# Update DE_info with SFARI and Neuronal information
genes_info = DE_info %>% mutate('ID'=rownames(.)) %>% left_join(SFARI_genes, by='ID') %>% 
  mutate(`gene-score`=ifelse(is.na(`gene-score`), 'None', `gene-score`)) %>%
  left_join(GO_neuronal, by='ID') %>% left_join(clusterings, by='ID') %>%
  mutate(Neuronal=ifelse(is.na(Neuronal), 0, Neuronal)) %>%
  mutate(gene.score=ifelse(`gene-score`=='None' & Neuronal==1, 'Neuronal', `gene-score`), 
         significant=padj<0.05 & !is.na(padj))


rm(DE_info, GO_annotations, clusterings)
```

---

# Exploratory analysis of clustering results

```{r}
print(paste0('Dynamic Tree leaves ', sum(genes_info$DynamicTree=='gray'), ' genes without cluster (', 
             round(mean(genes_info$DynamicTree=='gray')*100), '%)'))

print(paste0('Dynamic Hybrid leaves ', sum(genes_info$DynamicHybrid=='gray'), ' genes without cluster (', 
             round(mean(genes_info$DynamicHybrid=='gray')*100), '%)'))

```

Dynamic Tree leaves 31% of the genes without a cluster and comparing the Top modules of both clustering techniques (based on the dendrograms from 19_10_19_WGCNA_Rmd), the modules from this method are no longer cleaner than the modules belonging to the other method, so perhaps Dynamic Tree is not the best option with this new dataset...

There seems to be a relation between DE and module membership, being DE a more restrictive condition than being assigned to a cluster.

**This could be a consequence of the sva batch correction, because all the patterns that survived were diagnosis related, so it could have left all the genes with no relation to ASD without any recognisable pattern, which left them at the top of the dendrogram, which are the ones the clustering algorithm didn't assign to any module**
```{r pca_plots, warning=FALSE, fig.width=10}
pca = datExpr %>% prcomp

plot_data = data.frame('ID'=rownames(datExpr), 'PC1' = pca$x[,1], 'PC2' = pca$x[,2]) %>%
            left_join(genes_info, by='ID') %>% mutate('hasCluster'=DynamicTree!='gray', 
                                                      'hasSFARIScore'=`gene-score`!='None') %>%
            apply_labels(`gene-score`='SFARI Gene score', DynamicTree = 'Dynamic Tree Algorithm', 
                         significant = 'Differentially Expressed', hasCluster = 'Belongs to a Module',
                         hasSFARIScore = 'Has a SFARI Score', syndromic = 'Has syndromic tag')

p1 = plot_data %>% ggplot(aes(PC1, PC2, color=hasCluster)) + geom_point(alpha=0.2) + 
  theme_minimal() + ggtitle('Genes are assigned to a cluster') + theme(legend.position='bottom')

p2 = plot_data %>% ggplot(aes(PC1, PC2, color=significant)) + geom_point(alpha=0.2) + 
  theme_minimal() + ggtitle('Genes were found to be DE') + theme(legend.position='bottom')

grid.arrange(p1, p2, nrow=1)

rm(pca, p1, p2)
```

Most of the genes that don't have a cluster (97%) are not differentially expressed.
```{r}
cro(plot_data$significant, list(plot_data$hasCluster, total()))
```

Maybe it's not such a bad idea to keep only the genes that were assigned a module ...

187 of the SFARI genes (20%) are not assigned to any cluster
```{r}
cro(plot_data$hasSFARIScore, list(plot_data$hasCluster, total()))
```

We don't lose that many high ranking genes ...
```{r}
cro(plot_data$`gene-score`, list(plot_data$hasCluster, total()))

rm(plot_data)
```

## Conclusion:

I'm no longer sure about what to do. The main ndifference between algorithms is that Dynamic Hybrid clusters outlier genes and Dynamic Tree leaves them out... so Dynamic Tree would give me a 'cleaner' group of genes to work with, without losing many SFARI genes, but Dynamic Hybrid has less and more balanced clusters...

**Using the clustering from Dynamic Hybrid for now**
```{r select_clustering_method}
clustering_selected = 'DynamicHybridMergedSmall'
genes_info$Module = genes_info[,clustering_selected]
```

---

# Relate modules to external clinical traits

## Quantifying module-trait associations

In the WGCNA documentation they use Pearson correlation to calculate correlations, I think all of their variables were continuous. Since I have categorical variables I'm going to use the `hetcor` function, that calculates Pearson, polyserial or polychoric correlations depending on the type of variables involved.

- I'm not sure how the corPvalueStudent function calculates the p-values and I cannot find any documentation...

- Compared correlations using Pearson correlation and with hetcor and they are very similar, but a bit more extreme with hetcor. The same thing happens with the p-values.

```{r calc_module_trait_associations}
datTraits = datMeta %>% dplyr::select(Diagnosis_, Region, Sex, Age, PMI, RNAExtractionBatch) %>%
            rename('Diagnosis' = Diagnosis_, 'ExtractionBatch' = RNAExtractionBatch)

# Recalculate MEs with color labels
ME_object = datExpr %>% t %>% moduleEigengenes(colors = genes_info[,clustering_selected])
MEs = orderMEs(ME_object$eigengenes)

# Calculate correlation between eigengenes and the traits and their p-values
moduleTraitCor = MEs %>% apply(2, function(x) hetcor(x, datTraits, ML=TRUE)$correlations[1,-1]) %>% t
rownames(moduleTraitCor) = colnames(MEs)
colnames(moduleTraitCor) = colnames(datTraits)

# In case there are any NAs
if(sum(!complete.cases(moduleTraitCor))>0){
  print(paste0(sum(is.na(moduleTraitCor)),' correlation(s) could not be calculated')) 
  
  # Calculate the correlation tha failed with hetcor()
  missing_modules = rownames(moduleTraitCor)[is.na(moduleTraitCor[,1])]
  
  for(m in missing_modules){
    cat(paste0('Correcting Module-Diagnosis correlation for Module ', m))
    moduleTraitCor[m,'Diagnosis'] = polyserial(MEs[,m], datTraits$Diagnosis)
  }
}

# Calculate p-values
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nrow(datExpr))

# Create text matrix for the Heatmap
textMatrix = paste0(signif(moduleTraitCor, 2), ' (', signif(moduleTraitPvalue, 1), ')')
dim(textMatrix) = dim(moduleTraitCor)

rm(ME_object)
```

**Note:** The correlations between Modules and Diagonsis that cannot be calculated, weirdly enough, is because the initial correlation is too high, so it would be a very bad thing to lose these modules because of this numerical error. I’m going to fill in the values using the polyserial function, which doesn’t give exactly the same results as the hetcor() function, but it’s quite similar.


Modules have very strong correlations with Diagnosis and not much else, perhaps a little PMI
```{r plot_heatmap, fig.width=10, fig.height=10}
# Sort moduleTraitCor by Diagnosis
moduleTraitCor = moduleTraitCor[order(moduleTraitCor[,1], decreasing=TRUE),]
moduleTraitPvalue = moduleTraitPvalue[order(moduleTraitCor[,1], decreasing=TRUE),]

# Create text matrix for the Heatmap
textMatrix = paste0(signif(moduleTraitCor, 2), ' (', signif(moduleTraitPvalue, 1), ')')
dim(textMatrix) = dim(moduleTraitCor)


labeledHeatmap(Matrix = moduleTraitCor, xLabels = names(datTraits), yLabels =  gsub('ME','',rownames(moduleTraitCor)), 
               yColorWidth=0, colors = brewer.pal(11,'PiYG'), bg.lab.y = gsub('ME','',rownames(moduleTraitCor)),
               textMatrix = textMatrix, setStdMargins = FALSE, cex.text = 0.8, cex.lab.y = 0.75, zlim = c(-1,1),
               main = paste('Module-Trait relationships'))
```

```{r, warning=FALSE}
diagnosis_cor = data.frame('Module' = gsub('ME','',rownames(moduleTraitCor)),
                           'MTcor' = moduleTraitCor[,'Diagnosis'],
                           'MTpval' = moduleTraitPvalue[,'Diagnosis'])

genes_info = genes_info %>% left_join(diagnosis_cor, by='Module')

rm(moduleTraitCor, moduleTraitPvalue, datTraits, textMatrix, diagnosis_cor)
```

## Gene Significance and Module Membership

<br>

- **Gene significance:** absolute value between the correlation between the gene and the trait we are interested in

- **Module membership:** Correlation of the module eigengene and the gene expression profile

```{r gene_significance, warning=FALSE}
GS_info = data.frame('ID' = rownames(datExpr),
                     'GS' = datExpr %>% apply(1, function(x) hetcor(x, datMeta$Diagnosis_)$correlations[1,2])) %>%
          mutate('GSpval' = corPvalueStudent(GS, ncol(datExpr)))
```

```{r module_membership}
# It's more efficient to iterate the correlations one by one, otherwise it calculates correlations between the eigengenes and also between the genes, which we don't need

#setup parallel backend to use many processors
cores = detectCores()
  cl = makeCluster(cores-1)
registerDoParallel(cl)

# Create matrix with MM by gene
MM = foreach(i=1:nrow(datExpr), .combine=rbind) %dopar% {
  library(polycor)
  tempMatrix = apply(MEs, 2, function(x) hetcor(as.numeric(datExpr[i,]), x)$correlations[1,2])
  tempMatrix
}

# Stop clusters
stopCluster(cl)

rownames(MM) = rownames(datExpr)
colnames(MM) = paste0('MM',gsub('ME','',colnames(MEs)))

# Calculate p-values
MMpval = MM %>% corPvalueStudent(ncol(datExpr)) %>% as.data.frame
colnames(MMpval) = paste0('MMpval', gsub('ME','',colnames(MEs)))

rm(cores, cl)
```

--- 

# Create and save dataset

```{r create_dataset, warning=FALSE}
MM = MM %>% as.data.frame %>% mutate(ID = rownames(.))
MMpval = MMpval %>% as.data.frame %>% mutate(ID = rownames(.))

dataset = genes_info %>% dplyr::select(ID, `gene-score`, clustering_selected, MTcor, MTpval) %>%
          left_join(GS_info, by='ID') %>%
          left_join(MM, by='ID') %>%
          left_join(MMpval, by='ID')

write.csv(dataset, file = paste0('./../Data/Gandal/dataset_', clustering_selected, '.csv'), row.names = FALSE)
```
<br><br>

---

#### Session info

```{r print_session_info}
sessionInfo()
```
<br><br>